src/Playwright.NUnit/TimeoutCommand.cs

```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.Playwright.TestAdapter;
+using NUnit.Framework;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+using NUnit.Framework.Internal.Commands;
+
+namespace Microsoft.Playwright.NUnit;
+
+public class TimeoutCommand : BeforeTestCommand
+{
+    protected bool TearDownRan { get; set; }
+    internal TimeoutCommand(TestCommand innerCommand) : base(innerCommand) { }
+
+    public override TestResult Execute(TestExecutionContext context)
+    {
+        var timeout = PlaywrightSettingsProvider.TestTimeout;
+        if (timeout == null || Debugger.IsAttached)
+        {
+            return innerCommand.Execute(context);
+        }
+
+        if (PlaywrightTimeoutIsSmallerThanNUnitTimeout(context, timeout.Value, out var result))
+        {
+            return result!;
+        };
+
+        ExecuteWithTimeoutAsync(context, timeout.Value).GetAwaiter().GetResult();
+        return context.CurrentResult;
+    }
+
+    private async Task ExecuteWithTimeoutAsync(TestExecutionContext context, int timeout)
+    {
+        string timeoutExceptionMessage = $"Test exceeded Timeout value of {timeout}ms";
+        // 1. Race Test execution against the specified timeout.
+        var testExecutionTask = Task.Run(() => innerCommand.Execute(context));
+        var timeoutTask = Task.Delay(timeout);
+        await Task.WhenAny(testExecutionTask, timeoutTask);
+
+        // 2. If timeout was reached, call the tearDownHandlers and
+        // cancel by that the underlying BrowserContext calls.
+        if (timeoutTask.IsCompleted)
+        {
+            // 2.1 Call the test teardown handlers.
+            await RunSetUpOrTearDownMethod(context, isTearDown: true);
+            TearDownRan = true;
+
+            // 2.2 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
+        }
+
+        if (testExecutionTask.IsCompleted)
+        {
+            try
+            {
+                context.CurrentResult = await testExecutionTask;
+            }
+            catch (System.Exception exception)
+            {
+                context.CurrentResult = context.CurrentTest.MakeTestResult();
+                context.CurrentResult.RecordException(new AggregateException(new Exception[] { new TimeoutException(timeoutExceptionMessage), exception }), FailureSite.Test);
+            }
+            return;
+        }
+
+        context.CurrentResult = context.CurrentTest.MakeTestResult();
+        context.CurrentResult.RecordException(new TimeoutException(timeoutExceptionMessage));
+        return;
+    }
+
+    protected async Task RunSetUpOrTearDownMethod(TestExecutionContext context, bool isSetUp = false, bool isTearDown = false)
```

Comment:
I feel uncomfortable with all the internal details of NUnit we are touching here.